proxy/plugin/cache/base.py

```diff
@@ -27,34 +32,109 @@ class BaseCacheResponsesPlugin(HttpProxyBasePlugin):
     Different storage backends can be used per request if required.
     """
 
+    class EnabledDescriptor:
+        def __init__(self, enabled: bool = False) -> None:
+            self.enabled = multiprocessing.Event()
+            if enabled:
+                self.enabled.set()
+
+        def __get__(self, obj: Optional[object], owner: type) -> Any:
+            if obj is None:
+                return self.enabled
+            return None
+
+    # Dynamically enable / disable cache
+    enabled = EnabledDescriptor(True)
+
     def __init__(
             self,
             *args: Any,
             **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
         self.store: Optional[CacheStore] = None
+        self.__enabled: Optional[bool] = None
 
     def set_store(self, store: CacheStore) -> None:
         self.store = store
 
     def before_upstream_connection(
             self, request: HttpParser) -> Optional[HttpParser]:
+        self.__enabled = self.__class__.enabled.is_set()
+        if not self.__enabled:
+            return request
+
         assert self.store
+        logger.info("Upstream connexion %s:%d %s" %
+                    (text_(request.host), request.port if request.port else 0, text_(request.path)))
+
+        if request.port == 443:
```

Comment:
We can perform TLS handshake with the client without connecting with upstream.  See `utils/wrap_socket` utility here https://github.com/abhinavsingh/proxy.py/blob/develop/proxy/common/utils.py#L154-L166 . We'll have to re-use server side certificates generated previously during TLS interception for handshake with the client.  Wdyt?

Reply:
Yes, I already looked in this direction. This is what I do locally for automatic testing purposes. I pushed my work on a branch on my fork https://github.com/pasccom/proxy.py/tree/develop_cache_no_upstream
I did not create a PR yet because I wanted this one first to be fixed before going further. The major issue with not contacting upstream is that I ensure `do_connect` is falsy in https://github.com/abhinavsingh/proxy.py/blob/8cc349be488bd8fc5c23dea4f1f83c8d0c02a6f0/proxy/http/proxy/server.py#L337-L338 and consequently `self.server` is `None` at https://github.com/abhinavsingh/proxy.py/blob/8cc349be488bd8fc5c23dea4f1f83c8d0c02a6f0/proxy/http/proxy/server.py#L279

Reply:
Correct, `before_upstream_connection` callback can return `None` to avoid connecting to upstream.  Internally it will keep `do_connect = False`.

And thanks for bringing this to notice, `on_client_data` does ensures server connection before proceeding.  Might have to revisit the request state diagram to clean up some of these bits/expectations.